src/_zkapauthorizer/replicate.py

```diff
@@ -23,14 +23,77 @@
 ]
 
 from collections.abc import Awaitable
+from io import BytesIO
 from typing import BinaryIO, Callable
 
+import cbor2
+from attrs import frozen
 from twisted.python.lockfile import FilesystemLock
 
 from .config import REPLICA_RWCAP_BASENAME
 from .tahoe import Tahoe, attenuate_writecap
 
 
+@frozen
+class Change:
+    """
+    Represent an item in a replication event stream
+    """
+
+    sequence: int  # the sequence-number of this event
+    statement: str  # the SQL statement string
+
+
+@frozen
+class EventStream:
+    """
+    A series of database operations represented as `Change` instances.
+    """
+
+    changes: tuple[Change]
+
+    def highest_sequence(self):
```

Comment:
The design doc says that event stream objects have the lowest sequence number they contain in their metadata.  Maybe all of the logic works out with highest too?  Maybe the design doc could be clearer about this logic, too.

Specifically, the words in the design doc are:

> The sequence number of the first statement it includes is added as metadata for that object in the containing directory.

Please make the implementation and doc consistent (change whichever you think it is right to change).

Reply:
Ahh, hmmm ... okay. I think why I used "highest" is because the design wants to make decisions like "is it too long since that last one" and "does this snapshot contain everything in event-stream X" and both those want the maximum number. (I guess strictly speaking the second question needs _both_ numbers, but a "snapshot that contains sequence number X" is supposed to wrap-up every event with `sequences < X` right?)

Reply:
So, in that prototyping branch, event-streams are linked to the replica as `event-stream-<number>` which is I guess either the highest or lowest number in that event-stream (depending which I change).

If it's the lowest, pruning becomes harder. Let's say I have:

- a new snapshot with highest sequence number of 122
- on-replica event-stream files: `event-stream-0`,  `event-stream-123`

Can we delete `event-stream-0`?

We'd have to download it to be sure .. usually it would be the case with the above that the first file contains events 0 -> 122, but we don't really _know_ that. On the other hand, if that was `event-stream-122` and `event-stream-432` then we can tell just from the filenames/metadata (without downloading) that we can delete only `event-stream-122` right now (because a snapshop with "122" contains all the events to that point).